### PR TITLE
Implements Unique-Key possibility (issue #46)

### LINKED ISSUE #46: Unique-Key constraint configurable
There is not jet a possibility to create containers with a unique-key-constraint.

Because of that it would be great if one could specify those. Maybe one could add an attribute to the Property that should be unique, int the Item that implaments the IItem interface.

Then when a container is created, the type it is created for is checked for this attribute.

### PR DESCRIPTION
**Pull request by VeryCautious**
- Addes the UniqueKeyAttribute
- Addes IUniqueKeyPolicyProvider and a default implementation
- Addes Tests for the Provider
- Uses the policies when creating a container 

Fixes #46

### CODE REVIEW

File: Microsoft.Azure.CosmosRepository/src/Providers/DefaultCosmosUniqueKeyPolicyProvider.cs [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    class DefaultCosmosUniqueKeyPolicyProvider : ICosmosUniqueKeyPolicyProvider
+    {
+
+        static readonly Type _uniqueKeyAttributeType = typeof(UniqueKeyAttribute);
+        static readonly ConcurrentDictionary<Type, UniqueKeyPolicy> _uniqueKeyPolicyMap =
+            new ConcurrentDictionary<Type, UniqueKeyPolicy>();
+
+        /// <inheritdoc />
+        public UniqueKeyPolicy GetUniqueKeyPolicy<TItem>() where TItem : IItem =>
+            _uniqueKeyPolicyMap.GetOrAdd(typeof(TItem), GetPartitionKeyNameFactory);
```
Comment (IEvangelist) [inline Microsoft.Azure.CosmosRepository/src/Providers/DefaultCosmosUniqueKeyPolicyProvider.cs:24]:
```suggestion
            _uniqueKeyPolicyMap.GetOrAdd(typeof(TItem), GetUniqueKeyPolicyFactory);
```

Name this more appropriately, as it was copied and pasted from the `DefaultCosmosPartitionKeyPathProvider`.

File: Microsoft.Azure.CosmosRepository/src/Providers/DefaultCosmosUniqueKeyPolicyProvider.cs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+        static UniqueKeyPolicy GetPartitionKeyNameFactory(Type type)
```
Comment (IEvangelist) [inline Microsoft.Azure.CosmosRepository/src/Providers/DefaultCosmosUniqueKeyPolicyProvider.cs:26]:
```suggestion
        static UniqueKeyPolicy GetUniqueKeyPolicyFactory(Type type)
```

File: Microsoft.Azure.CosmosRepository/test/Providers/DefaultCosmosUniqueKeyPolicyProvider.cs [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    public class SomeInterestingClass2 : Item
+    {
+        [UniqueKey("addressKey")]
+        public string Street { get; set; } = "Street1";
+        [UniqueKey("addressKey")]
+        public string HouseNumber { get; set; } = "1";
+
+        public string Name { get; set; } = "John";
+
+        protected override string GetPartitionKeyValue() => Name;
```
Comment (IEvangelist) [inline Microsoft.Azure.CosmosRepository/test/Providers/DefaultCosmosUniqueKeyPolicyProvider.cs:77]:
```suggestion
```

No reason to `override` this as it's not part of the test case and adds a bit of confusion.

File: Microsoft.Azure.CosmosRepository/test/Providers/DefaultCosmosUniqueKeyPolicyProvider.cs [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    {
+        [UniqueKey("addressKey")]
+        public string Street { get; set; } = "Street1";
+        [UniqueKey("addressKey")]
+        public string HouseNumber { get; set; } = "1";
+
+        [UniqueKey("nameKey")]
+        public string Name { get; set; } = "John";
+
+        protected override string GetPartitionKeyValue() => Name;
```
Comment (IEvangelist) [inline Microsoft.Azure.CosmosRepository/test/Providers/DefaultCosmosUniqueKeyPolicyProvider.cs:90]:
```suggestion
```

File: Microsoft.Azure.CosmosRepository/test/Providers/DefaultCosmosUniqueKeyPolicyProvider.cs [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+
+    public class SomeInterestingClass : Item
+    {
+        public string Street { get; set; } = "Street1";
+        public string HouseNumber { get; set; } = "1";
+
+        [UniqueKey("nameKey")]
+        public string Name { get; set; } = "John";
+
+        protected override string GetPartitionKeyValue() => Name;
```
Comment (IEvangelist) [inline Microsoft.Azure.CosmosRepository/test/Providers/DefaultCosmosUniqueKeyPolicyProvider.cs:65]:
```suggestion
```

File: Microsoft.Azure.CosmosRepository/test/Providers/DefaultCosmosUniqueKeyPolicyProvider.cs [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+    public class DefaultCosmosUniqueKeyPolicyProviderTests
+    {
+        [Fact]
+        public void CosmosUniqueKeyPolicyProviderCorrectlyGetsOneUniqueKey()
+        {
+            ICosmosUniqueKeyPolicyProvider provider = new DefaultCosmosUniqueKeyPolicyProvider();
+
+            UniqueKeyPolicy policy  = provider.GetUniqueKeyPolicy<SomeInterestingClass>();
+            Assert.Equal("/Name", policy.UniqueKeys.Single().Paths.Single());
+
```
Comment (IEvangelist) [inline Microsoft.Azure.CosmosRepository/test/Providers/DefaultCosmosUniqueKeyPolicyProvider.cs:25]:
```suggestion
```

File: Microsoft.Azure.CosmosRepository/test/Providers/DefaultCosmosContainerProviderTests.cs [hunk 7]
```diff
@@ -53,6 +55,7 @@ public class DefaultCosmosContainerProviderTests
                     new TestCosmosClientProvider(),
                     new DefaultCosmosPartitionKeyPathProvider(),
                     null,
+                    new DefaultCosmosUniqueKeyPolicyProvider(),
```
Comment (IEvangelist) [inline Microsoft.Azure.CosmosRepository/test/Providers/DefaultCosmosContainerProviderTests.cs:58]: We need to add a test case to verify that passing a `null` as the `DefaultCosmosUniqueKeyPolicyProvider` argument throws, like the other tests here.

File: Microsoft.Azure.CosmosRepository/src/Providers/DefaultCosmosUniqueKeyPolicyProvider.cs [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+                foreach (PropertyInfo property in kv.Value)
+                {
+                    uniqueKey.Paths.Add("/" + property.Name);
+                }
+                policy.UniqueKeys.Add(uniqueKey);
+            }
+
+            return policy;
+        }
+
```
Comment (IEvangelist) [inline Microsoft.Azure.CosmosRepository/src/Providers/DefaultCosmosUniqueKeyPolicyProvider.cs:54]:
```suggestion
            Dictionary<string, List<PropertyInfo>> dict = new();
            IEnumerable<PropertyInfo> uniqueKeyed = type.GetProperties().Where((p) => Attribute.IsDefined(p, _uniqueKeyAttributeType));

            foreach (PropertyInfo property in uniqueKeyed)
            {
                string keyName = property.GetCustomAttribute<UniqueKeyAttribute>().KeyName;
                if (!dict.ContainsKey(keyName))
                {
                    dict.Add(keyName, new List<PropertyInfo>());
                }
                dict[keyName].Add(property);
            }

            UniqueKeyPolicy policy = new();
            foreach (KeyValuePair<string, List<PropertyInfo>> kv in dict)
            {
                UniqueKey uniqueKey = new();
                foreach (PropertyInfo property in kv.Value)
                {
                    uniqueKey.Paths.Add($"/{property.Name}");
                }
                policy.UniqueKeys.Add(uniqueKey);
            }

            return policy;
        }
```

I think before we ship this - I'll want to run a few more tests on it. It seems like this is something that could be optimized a bit more as well. Also, shouldn't be possible for there to be no policy when there are no attributes used on an item?

File: Microsoft.Azure.CosmosRepository/src/Attributes/UniqueKeyAttribute.cs [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+    [AttributeUsage(AttributeTargets.Property, AllowMultiple = true, Inherited = true)]
+    class UniqueKeyAttribute : Attribute
+    {
+        public string KeyName { get; } = "uniqueKey";
+
+        public UniqueKeyAttribute(string keyName) =>
+            KeyName = keyName ?? throw new ArgumentNullException(nameof(keyName), "A keyname is required.");
+
+    }
+
```
Comment (IEvangelist) [inline Microsoft.Azure.CosmosRepository/src/Attributes/UniqueKeyAttribute.cs:17]:
```suggestion
    }
```

Remove extra new lines.

**Review by IEvangelist — CHANGES_REQUESTED** [review]
I like the direction this is heading, and thank you for putting this together. Here are some thoughts on improving this PR.

File: Microsoft.Azure.CosmosRepository/src/Providers/DefaultCosmosUniqueKeyPolicyProvider.cs (hunk 8 above)
Comment (VeryCautious) [reply Microsoft.Azure.CosmosRepository/src/Providers/DefaultCosmosUniqueKeyPolicyProvider.cs:53]: That is true. As is now I think it would implicitly do so by providing a policy with no keys. But maybe it would be better to return either null or maybe check if there are any unique keys to beginn with.

**Comment by IEvangelist** [comment]
@all-contributors please add @VeryCautious for code and tests

File: Microsoft.Azure.CosmosRepository/test/Providers/DefaultCosmosContainerProviderTests.cs (hunk 7 above)
Comment (VeryCautious) [reply Microsoft.Azure.CosmosRepository/test/Providers/DefaultCosmosContainerProviderTests.cs:58]: added that in commit bb520e3 ("addes tests")

File: Microsoft.Azure.CosmosRepository/src/Extensions/ServiceCollectionExtensions.cs [hunk 10]
```diff
@@ -43,6 +43,7 @@ public static class ServiceCollectionExtensions
                     .AddSingleton(typeof(ICosmosContainerProvider<>), typeof(DefaultCosmosContainerProvider<>))
                     .AddSingleton<ICosmosPartitionKeyPathProvider, DefaultCosmosPartitionKeyPathProvider>()
                     .AddSingleton<ICosmosContainerNameProvider, DefaultCosmosContainerNameProvider>()
+                    .AddSingleton<ICosmosUniqueKeyPolicyProvider,DefaultCosmosUniqueKeyPolicyProvider>()
```
Comment (IEvangelist) [inline Microsoft.Azure.CosmosRepository/src/Extensions/ServiceCollectionExtensions.cs:46]:
```suggestion
                    .AddSingleton<ICosmosUniqueKeyPolicyProvider, DefaultCosmosUniqueKeyPolicyProvider>()

```

File: Microsoft.Azure.CosmosRepository/src/Providers/DefaultCosmosUniqueKeyPolicyProvider.cs [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+using System.Text;
+using Microsoft.Azure.Cosmos;
+using Microsoft.Azure.CosmosRepository.Attributes;
+
+namespace Microsoft.Azure.CosmosRepository.Providers
+{
+    class DefaultCosmosUniqueKeyPolicyProvider : ICosmosUniqueKeyPolicyProvider
+    {
+        static readonly Type _uniqueKeyAttributeType = typeof(UniqueKeyAttribute);
+
```
Comment (IEvangelist) [inline Microsoft.Azure.CosmosRepository/src/Providers/DefaultCosmosUniqueKeyPolicyProvider.cs:18]:
```suggestion
```

File: Microsoft.Azure.CosmosRepository/src/Providers/DefaultCosmosUniqueKeyPolicyProvider.cs [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+        static readonly ConcurrentDictionary<Type, UniqueKeyPolicy> _uniqueKeyPolicyMap = new();
```
Comment (IEvangelist) [inline Microsoft.Azure.CosmosRepository/src/Providers/DefaultCosmosUniqueKeyPolicyProvider.cs:19]:
```suggestion
        static readonly Type _uniqueKeyAttributeType = typeof(UniqueKeyAttribute);
        static readonly ConcurrentDictionary<Type, UniqueKeyPolicy> _uniqueKeyPolicyMap = new();
```

Comment (IEvangelist) [inline Microsoft.Azure.CosmosRepository/src/Providers/DefaultCosmosUniqueKeyPolicyProvider.cs:19]:
```suggestion
        static readonly Type _uniqueKeyAttributeType = typeof(UniqueKeyAttribute);
        static readonly ConcurrentDictionary<Type, UniqueKeyPolicy> _uniqueKeyPolicyMap = new();
```

**Review by IEvangelist — APPROVED** [review]
Looks good to me, we'll :shipit: - thank you 🙏
